File: chrome/src/java/org/openqa/selenium/chrome/ChromeBinary.java
package org.openqa.selenium.chrome;

import org.openqa.selenium.Platform;
import org.openqa.selenium.Proxy;
import static org.openqa.selenium.Proxy.ProxyType;
import org.openqa.selenium.WebDriverException;
import org.openqa.selenium.remote.internal.CircularOutputStream;
import org.openqa.selenium.remote.internal.SubProcess;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.ServerSocket;
import java.util.List;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.collect.Lists;

public class ChromeBinary {

  private static final String CHROME_LOG_FILE_PROPERTY = "webdriver.chrome.logFile";
  private static final int BACKOFF_INTERVAL = 2500;

  private volatile int linearBackoffCoefficient = 1;

  private final ChromeProfile profile;
  private final ChromeExtension extension;
  private final int port;
  private final SubProcess chromeProcess;

  protected String chromeBinaryLocation = null;

  /**
   * @param profile The Chrome profile to use.
   * @param extension The extension to launch Chrome with.
   * @throws WebDriverException If an error occurs locating the Chrome executable.
   * @see ChromeBinary(ChromeProfile, ChromeExtension, int)
   */
  public ChromeBinary(ChromeProfile profile, ChromeExtension extension) {
    this(profile, extension, 0);
  }

  /**
   * Creates a new instance for managing an instance of Chrome using the given
   * {@code profile} and {@code extension}.
   *
   * @param profile The Chrome profile to use.
   * @param extension The extension to launch Chrome with.
   * @param port Which port to start Chrome on, or 0 for any free port.
   * @throws WebDriverException If an error occurs locating the Chrome executable.
   */
  public ChromeBinary(ChromeProfile profile, ChromeExtension extension, int port) {
    this.profile = profile;
    this.extension = extension;
    this.port = port == 0 ? findFreePort() : port;
    String serverUrl = String.format("http://localhost:%d/chromeCommandExecutor", this.port);

    ProcessBuilder builder;
    try {
      List<String> commandline = getCommandline(serverUrl);
      builder = new ProcessBuilder(commandline);
    } catch (IOException e) {
      throw new WebDriverException(e);
    }

    File logFile = getLogFile();
    this.chromeProcess = logFile == null
        ? new SubProcess(builder)
        : new SubProcess(builder, new CircularOutputStream(logFile));
  }

  private static File getLogFile() {
    String logFile = System.getProperty(CHROME_LOG_FILE_PROPERTY);
    return logFile == null ? null : new File(logFile);
  }

  private static int findFreePort() {
    ServerSocket serverSocket = null;
    try {
      serverSocket = new ServerSocket(0);
      return serverSocket.getLocalPort();
    } catch (IOException e) {
      throw new WebDriverException(e);
    } finally {
      if (serverSocket != null) {
        try {
          serverSocket.close();
        } catch (IOException ignored) {
          // Oh well
        }
      }
    }
  }

  public ChromeProfile getProfile() {
    return profile;
  }

  public ChromeExtension getExtension() {
    return extension;
  }

  public int getPort() {
    return port;
  }

  /**
   * Starts the Chrome process for WebDriver.
   */
  public void start() {
    chromeProcess.launch();
    try {
      Thread.sleep(BACKOFF_INTERVAL * linearBackoffCoefficient);
    } catch (InterruptedException e) {
      //Nothing sane to do here
    }
  }

  @VisibleForTesting List<String> getCommandline(String serverUrl) throws IOException {
    List<String> commandline = Lists.newArrayList(Lists.newArrayList(
        getChromeBinaryLocation(),
        "--user-data-dir=" + profile.getDirectory().getAbsolutePath(),
        "--load-extension=" + extension.getDirectory().getAbsolutePath(),
        "--activate-on-launch",
        "--homepage=about:blank",
        "--no-first-run",
        "--disable-hang-monitor",
        "--disable-popup-blocking",
        "--disable-prompt-on-repost",
        "--no-default-browser-check"
    ));
    appendProxyArguments(commandline)
        .add(serverUrl);
    return commandline;
  }

  private List<String> appendProxyArguments(List<String> commandline) {
    Proxy proxy = profile.getProxy();
    if (proxy == null) {
      return commandline;
    }
    if (proxy.getProxyAutoconfigUrl() != null) {
      commandline.add("--proxy-pac-url=" + proxy.getProxyAutoconfigUrl());
    } else if (proxy.getHttpProxy() != null) {
      commandline.add("--proxy-server=" + proxy.getHttpProxy());
    } else if (proxy.isAutodetect()) {
      commandline.add("--proxy-auto-detect");
    } else if (proxy.getProxyType() == ProxyType.DIRECT) {
      commandline.add("--no-proxy-server");
    } else if (proxy.getProxyType() != ProxyType.SYSTEM) {
      throw new IllegalStateException("Unsupported proxy setting");
    }
    return commandline;
  }

  /**
   * @return Whether the Chrome process managed by this instance is still
   *     running.
   */
  public boolean isRunning() {
    return chromeProcess.isRunning();
  }

  /**
   * Kills the Chrome process managed by this instance.
   */
  public void kill() {
    chromeProcess.shutdown();
  }

  public void incrementBackoffBy(int diff) {
    linearBackoffCoefficient += diff;
  }

  /**
   * Locates the Chrome executable on the current platform.
   * First looks in the webdriver.chrome.bin property, then searches
   * through the default expected locations.
   * @return chrome.exe
   * @throws IOException if file could not be found/accessed
   */
  protected String getChromeBinaryLocation() throws IOException {
    if (!isChromeBinaryLocationKnown()) {
      chromeBinaryLocation = System.getProperty("webdriver.chrome.bin");
      if (chromeBinaryLocation == null) {
        if (Platform.getCurrent().is(Platform.WINDOWS)) {
          chromeBinaryLocation = getWindowsBinaryLocationFromRegistry();
          if (chromeBinaryLocation == null) {
            chromeBinaryLocation = getDefaultWindowsBinaryLocation();
          }
        } else if (Platform.getCurrent().is(Platform.UNIX)) {
          chromeBinaryLocation = "/usr/bin/google-chrome";
        } else if (Platform.getCurrent().is(Platform.MAC)) {
          String[] paths = new String[] {
            "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome",
            "/Users/" + System.getProperty("user.name") +
                "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome"};
          for (String path : paths) {
            File binary = new File(path);
            if (binary.exists()) {
              chromeBinaryLocation = binary.getCanonicalFile().getAbsoluteFile().toString();
              break;
            }
          }
        } else {
          throw new WebDriverException("Unsupported operating system.  " +
              "Could not locate Chrome.  Set webdriver.chrome.bin");
        }
      }
      if (!isChromeBinaryLocationKnown()) {
        throw new WebDriverException("Couldn't locate Chrome.  " +
            "Set webdriver.chrome.bin");
      }
    }
    return chromeBinaryLocation;
  }

  protected boolean isChromeBinaryLocationKnown() {
    return chromeBinaryLocation != null && new File(chromeBinaryLocation).exists();
  }

  /**
   * Returns null if couldn't read value from registry
   */
  protected static final String getWindowsBinaryLocationFromRegistry() {
    //TODO: Promote org.openqa.selenium.server.browserlaunchers.WindowsUtils
    //to common and reuse that to read the registry
    if (!Platform.WINDOWS.is(Platform.getCurrent())) {
      throw new UnsupportedOperationException("Cannot get registry value on non-Windows systems");
    }
    try {
      Process process = Runtime.getRuntime().exec(
          "reg query \"HKEY_LOCAL_MACHINE\\SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\App Paths\\chrome.exe\" /v \"\"");
      BufferedReader reader = new BufferedReader(new InputStreamReader(
          process.getInputStream()));
      process.waitFor();
      String line;
      while ((line = reader.readLine()) != null) {
        if (line.contains("    ")) {
          String[] tokens = line.split("REG_SZ");
          return tokens[tokens.length - 1].trim();
        }
      }
    } catch (IOException e) {
      //Drop through to return null
    } catch (InterruptedException e) {
      //Drop through to return null
    }
    return null;
  }

  protected static final String getDefaultWindowsBinaryLocation() {
    StringBuilder path = new StringBuilder();
    path.append(System.getProperty("user.home"));
    //XXX: Not localised for other languages
    if (Platform.VISTA.is(Platform.getCurrent())) {
      path.append("\\AppData\\Local");
    } else if (Platform.XP.is(Platform.getCurrent())) {
      path.append("\\Local Settings\\Application Data");
    }
    path.append("\\Google\\Chrome\\Application\\chrome.exe");
    return path.toString();
  }
}

File: chrome/src/csharp/webdriver-chrome/ChromeBinary.cs
using System;
using System.Diagnostics;
using System.Globalization;
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Threading;
using Microsoft.Win32;

namespace OpenQA.Selenium.Chrome
{
    /// <summary>
    /// Provides a mechanism to find the Chrome Binary
    /// </summary>
    internal class ChromeBinary
    {
        private static readonly string[] chromePaths = new string[]
        {
            "/usr/bin/google-chrome",
            "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome",
            string.Concat("/Users/", Environment.UserName, "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome")
        };

        private const int ShutdownWaitInterval = 2000;
        private const int StartWaitInterval = 2500;
        private static int linearStartWaitCoefficient = 1;
        private static string chromeFile = string.Empty;
        private ChromeProfile profile;
        private ChromeExtension extension;
        private Process chromeProcess;
        private int listeningPort;

        /// <summary>
        /// Initializes a new instance of the ChromeBinary class using the given <see cref="ChromeProfile"/> and <see cref="ChromeExtension"/>.
        /// </summary>
        /// <param name="profile">The Chrome profile to use.</param>
        /// <param name="extension">The extension to launch Chrome with.</param>
        internal ChromeBinary(ChromeProfile profile, ChromeExtension extension)
            : this(profile, extension, 0)
        {
        }

        /// <summary>
        /// Initializes a new instance of the ChromeBinary class using the given <see cref="ChromeProfile"/>, <see cref="ChromeExtension"/> and port value.
        /// </summary>
        /// <param name="profile">The Chrome profile to use.</param>
        /// <param name="extension">The extension to launch Chrome with.</param>
        /// <param name="port">The port on which to listen for commands.</param>
        internal ChromeBinary(ChromeProfile profile, ChromeExtension extension, int port)
        {
            this.profile = profile;
            this.extension = extension;
            if (port == 0)
            {
                FindFreePort();
            }
            else
            {
                this.listeningPort = port;
            }
        }

        /// <summary>
        /// Gets the port number on which the <see cref="ChromeExtension"/> should listen for commands.
        /// </summary>
        public int Port
        {
            get { return listeningPort; }
        }

        private static string ChromePathFromRegistry
        {
            get
            {
                return Registry.LocalMachine.OpenSubKey(
                                    "SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\App Paths\\").GetValue("chrome.exe").
                                    ToString();
            }
        }

        private string Arguments
        {
            get
            {
                return string.Concat(" --user-data-dir=\"", profile.ProfileDirectory, "\"", " --load-extension=\"", extension.ExtensionDirectory, "\"", " --activate-on-launch", " --homepage=about:blank", " --no-first-run", " --disable-hang-monitor", " --disable-popup-blocking", " --disable-prompt-on-repost", " --no-default-browser-check ");
            }
        }

        /// <summary>
        /// Increases the wait time used for starting the Chrome process.
        /// </summary>
        /// <param name="diff">Interval by which to increase the wait time.</param>
        public static void IncrementStartWaitInterval(int diff)
        {
            linearStartWaitCoefficient += diff;
        }

        /// <summary>
        ///  Starts the Chrome process for WebDriver. Assumes the passed directories exist.
        /// </summary>
        /// <exception cref="WebDriverException">When it can't launch will throw an error</exception>
        public void Start()
        {
            try
            {
                chromeProcess = Process.Start(new ProcessStartInfo(GetChromeFile(), string.Concat(Arguments, string.Format(CultureInfo.InvariantCulture, "http://localhost:{0}/chromeCommandExecutor", listeningPort))) { UseShellExecute = false });
            }
            catch (IOException e)
            { // TODO(AndreNogueira): Check exception type thrown when process.start fails
                throw new WebDriverException("Could not start Chrome process", e);
            }

            Thread.Sleep(StartWaitInterval * linearStartWaitCoefficient);
        }

        /// <summary>
        /// Kills off the Browser Instance
        /// </summary>
        public void Kill()
        {
            if ((chromeProcess != null) && !chromeProcess.HasExited)
            {
                // Ask nicely to close.
                chromeProcess.CloseMainWindow();
                chromeProcess.WaitForExit(ShutdownWaitInterval);

                // If it still hasn't closed, be rude.
                if (!chromeProcess.HasExited)
                {
                    chromeProcess.Kill();
                }

                chromeProcess = null;
                DeleteProfileDirectory(profile.ProfileDirectory);
            }
        }

        /// <summary>
        /// Locates the Chrome executable on the current platform. First looks in the webdriver.chrome.bin property, then searches
        /// through the default expected locations.
        /// </summary>
        /// <returns>chrome.exe path</returns>
        private static string GetChromeFile()
        {
            if (!IsChromeBinaryLocationKnown())
            {
                string chromeFileSystemProperty = Environment.GetEnvironmentVariable("webdriver.chrome.bin");
                if (chromeFileSystemProperty != null)
                {
                    chromeFile = chromeFileSystemProperty;
                }
                else
                {
                    if (Platform.CurrentPlatform.IsPlatformType(PlatformType.XP) || Platform.CurrentPlatform.IsPlatformType(PlatformType.Vista))
                    {
                        try
                        {
                            chromeFile = ChromePathFromRegistry;
                        }
                        catch (NullReferenceException)
                        {
                            if (Platform.CurrentPlatform.IsPlatformType(PlatformType.XP))
                            {
                                chromeFile =
                                    Path.Combine(
                                        Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
                                        "Google\\Chrome\\Application\\chrome.exe");
                            }
                            else
                            {
                                chromeFile = Path.Combine(Path.GetTempPath(), "..\\Google\\Chrome\\Application\\chrome.exe");
                            }
                        }
                    }
                    else if (Platform.CurrentPlatform.IsPlatformType(PlatformType.Unix))
                    {
                        // Thanks to a bug in Mono Mac and Linux will be treated the same  https://bugzilla.novell.com/show_bug.cgi?id=515570 but adding this in case
                        string chromeFileString = string.Empty;
                        bool foundPath = false;
                        foreach (string path in chromePaths)
                        {
                            FileInfo binary = new FileInfo(path);
                            if (binary.Exists)
                            {
                                chromeFileString = binary.FullName;
                                foundPath = true;
                                break;
                            }
                        }

                        if (!foundPath)
                        {
                            throw new WebDriverException("Couldn't locate Chrome. Set webdriver.chrome.bin");
                        }

                        chromeFile = chromeFileString;
                    }
                    else
                    {
                        throw new WebDriverException(
                            "Unsupported operating system. Could not locate Chrome.  Set webdriver.chrome.bin");
                    }
                }
            }

            return chromeFile;
        }

        private static bool IsChromeBinaryLocationKnown()
        {
            return !string.IsNullOrEmpty(chromeFile) && File.Exists(chromeFile);
        }

        private static void DeleteProfileDirectory(string directoryToDelete)
        {
            int numberOfRetries = 0;
            while (Directory.Exists(directoryToDelete) && numberOfRetries < 10)
            {
                try
                {
                    Directory.Delete(directoryToDelete, true);
                }
                catch (IOException)
                {
                    // If we hit an exception (like file still in use), wait a half second
                    // and try again. If we still hit an exception, go ahead and let it through.
                    System.Threading.Thread.Sleep(500);
                }
                catch (UnauthorizedAccessException)
                {
                    // If we hit an exception (like file still in use), wait a half second
                    // and try again. If we still hit an exception, go ahead and let it through.
                    System.Threading.Thread.Sleep(500);
                }
                finally
                {
                    numberOfRetries++;
                }

                if (Directory.Exists(directoryToDelete))
                {
                    Console.WriteLine("Unable to delete profile directory '{0}'", directoryToDelete);
                }
            }
        } 

        private void FindFreePort()
        {
            // Locate a free port on the local machine by binding a socket to
            // an IPEndPoint using IPAddress.Any and port 0. The socket will
            // select a free port.
            Socket portSocket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
            IPEndPoint socketEndPoint = new IPEndPoint(IPAddress.Any, 0);
            portSocket.Bind(socketEndPoint);
            socketEndPoint = (IPEndPoint)portSocket.LocalEndPoint;
            listeningPort = socketEndPoint.Port;
            portSocket.Close();
        }
    }
}
